kmod/src/forest.c

```diff
@@ -605,14 +662,282 @@ int scoutfs_forest_setup(struct super_block *sb)
+
+	finf->orphan_list_present_hint = true;
+}
+
+/*
+ * At the start of the transaction, alloc and mark dirty a meta block in
+ * anticipation of updating the OINO BF at the end of the transaction.
+ */
+int scoutfs_forest_oino_newtrans_alloc(struct super_block *sb, u64 trans_seq)
+{
+	struct scoutfs_block *new_bl = NULL;
+	struct scoutfs_block_ref *ref;
+	DECLARE_FOREST_INFO(sb, finf);
+	int ret;
+
+	WARN_ON_ONCE(finf->oino_bl);
+
+	ref = &finf->our_log.oino_bloom_ref;
+
+	/* free previous oino block, if any */
+	if (ref->blkno) {
+		ret = scoutfs_free_meta(sb, finf->alloc, finf->wri,
+					le64_to_cpu(ref->blkno));
+		BUG_ON(ret);
+		memset(ref, 0, sizeof(struct scoutfs_block_ref));
+	}
+
+	ret = scoutfs_block_dirty_ref(sb, finf->alloc, finf->wri, ref, SCOUTFS_BLOCK_MAGIC_BLOOM,
+				      &new_bl, 0, NULL);
+	if (ret < 0)
+		goto out;
+
+	finf->oino_bl = new_bl;
+
+out:
+	return ret;
+}
+
+int scoutfs_forest_oino_update(struct super_block *sb)
+{
+	struct scoutfs_bloom_block *bb;
+	DECLARE_FOREST_INFO(sb, finf);
+	u64 last_seq;
+	int ret;
+	int i;
+
+	if (!finf->oino_bl)
+		return 0;
+
+	bb = finf->oino_bl->data;
+
+	ret = scoutfs_client_get_last_seq(sb, &last_seq);
+	if (ret)
+		return ret;
+
+	bb->last_seq = cpu_to_le64(last_seq);
+	finf->our_log.oino_bloom_ref.seq = bb->hdr.seq;
+
+	/* Turn the counted BF into a classic BF */
+	for (i = 0; i < SCOUTFS_FOREST_BLOOM_BITS; i++) {
+		if (atomic_read(&finf->oino_counted_bloom[i]))
+			set_bit_le(i, bb->bits);
+	}
+	scoutfs_block_put(sb, finf->oino_bl);
+	finf->oino_bl = NULL;
+
+	if (!finf->shutdown)
+		schedule_delayed_work(&finf->oino_work, msecs_to_jiffies(OINO_WORK_DELAY_MS));
+
+	return 0;
+}
+
+#define MAX_ORPHAN_DELETE_BATCH 4096
+
+void scoutfs_oino_delete_work(struct work_struct *work)
+{
+	struct forest_info *finf = container_of(work, struct forest_info,
+						oino_work.work);
+	struct super_block *sb = finf->sb;
+	struct scoutfs_net_roots roots;
+	SCOUTFS_BTREE_ITEM_REF(iref);
+	struct scoutfs_log_trees lt = {{{ 0 }}};
+	struct scoutfs_bloom_block *bb;
+	struct scoutfs_bloom_block *tmp_bb = NULL;
+	struct scoutfs_block *bl;
+	struct scoutfs_key ltk;
+	int ret = 0;
+
+	scoutfs_inc_counter(sb, forest_oino_delete_work);
+
+	ret = scoutfs_client_get_roots(sb, &roots);
+	if (ret)
+		goto out;
+
+	scoutfs_key_init_log_trees(&ltk, 0, 0);
+	for (;; scoutfs_key_inc(&ltk)) {
+		ret = scoutfs_btree_next(sb, &roots.logs_root, &ltk, &iref);
+		if (ret == 0) {
+			if (iref.val_len == sizeof(lt)) {
+				ltk = *iref.key;
+				memcpy(&lt, iref.val, sizeof(lt));
+			} else {
+				ret = -EIO;
+			}
+			scoutfs_btree_put_iref(&iref);
+		}
+		if (ret < 0) {
+			if (ret == -ENOENT)
+				break;
+			goto out; /* including stale */
+		}
+
+		if (lt.oino_bloom_ref.blkno == 0)
+			continue;
+
+		bl = read_bloom_ref(sb, &lt.oino_bloom_ref);
+		if (IS_ERR(bl)) {
+			ret = PTR_ERR(bl);
+			goto out;
+		}
+		bb = bl->data;
+
+		if (!tmp_bb) {
+			tmp_bb = vmalloc(SCOUTFS_BLOCK_LG_SIZE);
```

Comment:
We should just unconditionally allocate this up at the top of the function.  Then we don't need to test it when trying to delete, and it won't add a leak of the bl ref with the goto out without the put.  And you don't need to memset the block to 0, you can use the vmalloc that takes flags and ask it to zero.. something like __vmalloc(bytes, GFP_NOFS | __GFP_ZERO, PAGE_KERNEL);   (oh, which also reminds me that this probably wants GFP_NOFS so it can't recurse back into the fs :))

Reply:
ok cool. also using GFP_ZERO on the vmalloced array of atomics, that also seems kosher.